app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,161 @@
+
+            val recentHistoryGroups = async {
+                historyMetadataStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                    .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
+                    .groupBy { it.key.searchTerm!! }
+                    .mapValues { group ->
+                        // Within a group, we dedupe entries based on their url so we don't display
+                        // a page multiple times in the same group, and we sum up the total view time
+                        // of deduped entries while making sure to keep the latest updatedAt value.
+                        val metadataInGroup = group.value
+                        val metadataUrlGroups = metadataInGroup.groupBy { metadata -> metadata.key.url }
+                        metadataUrlGroups.map { metadata ->
+                            metadata.value.reduce { acc, elem ->
+                                acc.copy(
+                                    totalViewTime = acc.totalViewTime + elem.totalViewTime,
+                                    updatedAt = max(acc.updatedAt, elem.updatedAt)
+                                )
+                            }
+                        }
+                    }
+                    .toList()
+            }
+
+            updateState(historyHighlights.await(), recentHistoryGroups.await())
+        }
+    }
+
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ): List<RecentlyVisitedItem> {
+        val dedupedHistoryHighlights = dedupeVisits(historyHighlights, historyGroups)
+        val recentHighlights = dedupedHistoryHighlights
+            .map {
+                RecentHistoryHighlight(
+                    title = if (it.title.isNullOrBlank()) it.url else it.title!!,
+                    url = it.url
+                )
+            }
+        val recentGroups = historyGroups
+            .map { (title, data) ->
+                RecentHistoryGroup(
+                    title = title,
+                    historyMetadata = data
+                )
+            }
+            .sortedByDescending { it.lastUpdated() }
+            .take(MAX_RESULTS_TOTAL)
+
+        val totalItemsCount = recentHighlights.size + recentGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
+            recentGroups + recentHighlights
+        } else {
+            var groupsCount = 0
+            var highlightCount = 0
+            while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) {
```

Comment:
⚠️ There's a risk here that this loops infinitely. If both conditions aren't met we never increment any counter.

Reply:
Looked more at this.
I think the first check from the `if` block will guarantee we have enough items for the counters which just count this items of both types to be incremented up to the `MAX_RESULTS_TOTAL`
```
val totalItemsCount = recentHighlights.size + recentGroups.size

return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
    recentGroups + recentHighlights
} else {
    var groupsCount = 0
    var highlightCount = 0
    while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) { .. }
```
In the `while` block I'm using `recentHighlights.getOrNull(highlightCount)` to increment the counters starting from position 0 and if I know I have at least `MAX_RESULTS_TOTAL` items available then that `getOrNull` check has to return a real instance at least for one type up to that total we want.

Not sure what the edgecase would be.

Reply:
Ah yes :). I missed that this is in the else block where it's guaranteed that we have at least MAX_RESULTS_TOTAL results.